internal/idtokenvalidator/idtokenvalidator.go

```diff
@@ -0,0 +1,196 @@
+package idtokenvalidator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+// OptionalVerification allows validating optional claims that might not always be in the ID token.
+type OptionalVerification struct {
+	MaxAge       time.Duration
+	Nonce        string
+	Organization string
+}
+
+// IDTokenValidator is used to validate ID tokens retrieved from Auth0.
+type IDTokenValidator struct {
+	alg            jwa.SignatureAlgorithm
+	audience       string
+	clientSecret   []byte
+	clockTolerance time.Duration
+	httpClient     *http.Client
+	issuer         string
+	jwks           jwk.Cache
+	jwksURL        string
+}
+
+// New creates and returns a new IDTokenValidator.
+func New(
+	ctx context.Context,
+	domain string,
+	clientID string,
+	clientSecret string,
+	idTokenSigningAlg string,
+	opts ...Option,
+) (*IDTokenValidator, error) {
+	if i := strings.Index(domain, "//"); i != -1 {
+		domain = domain[i+2:]
+	}
+	alg, err := determineAlg(idTokenSigningAlg)
+	if err != nil {
+		return nil, err
+	}
+
+	i := &IDTokenValidator{
+		clientSecret:   []byte(clientSecret),
+		alg:            alg,
+		clockTolerance: time.Minute,
+		issuer:         "https://" + domain + "/",
+		audience:       clientID,
+	}
+
+	for _, option := range opts {
+		option(i)
+	}
+
+	if alg == jwa.RS256 {
+		i.jwksURL = i.issuer + ".well-known/jwks.json"
+		i.jwks = *jwk.NewCache(ctx)
+		registerOpts := []jwk.RegisterOption{}
+		if i.httpClient != nil {
+			registerOpts = append(registerOpts, jwk.WithHTTPClient(i.httpClient))
+		}
+
+		err = i.jwks.Register(i.jwksURL, registerOpts...)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = i.jwks.Refresh(ctx, i.jwksURL)
```

Comment:
Should we always force refresh the JWKS, or should we rely on the cache expiration?

Reply:
This refresh is so that we can ensure the JWKS is available, then the jwx library should handle refreshing at least every 15 minutes from my research.

I think before GA it might be good to ensure that the behaviour here is consistent with other SDKs (I think node-auth0 has a refresh time of 10 minutes for example)

Reply:
My bad, it's fine 🤦🏼‍♀️